File: cses/06_tree_algorithms/1138_path-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    q = int(data[pos]); pos += 1

    val = [0] * (n + 1)
    for i in range(1, n + 1):
        val[i] = int(data[pos]); pos += 1

    # Danh sách kề
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(data[pos]); pos += 1
        b = int(data[pos]); pos += 1
        adj[a].append(b)
        adj[b].append(a)

    # Euler tour bằng DFS lặp (tránh RecursionError khi n lớn): tính tin[], tout[], parent[].
    # Dấu âm -v được dùng làm marker để chốt tout khi rời khỏi đỉnh v.
    tin = [0] * (n + 1)
    tout = [0] * (n + 1)
    parent = [0] * (n + 1)
    timer = 0
    stack = [1]
    while stack:
        v = stack.pop()
        if v > 0:
            timer += 1
            tin[v] = timer  # Lúc vào đỉnh v
            stack.append(-v)  # Đẩy marker để gán tout khi ra khỏi v
            pv = parent[v]
            for u in adj[v]:
                if u != pv:  # Bỏ qua cạnh quay về cha
                    parent[u] = v
                    stack.append(u)
        else:
            v = -v
            tout[v] = timer  # Lúc ra khỏi đỉnh v

    # BIT có chỉ số 1..size, size = n+1 (vì có thể cộng tại vị trí tout[v]+1 = n+1)
    size = n + 1

    # Dựng mảng hiệu (difference array): mỗi đỉnh cộng +val tại tin, -val tại tout+1.
    diff = [0] * (size + 2)
    for v in range(1, n + 1):
        vv = val[v]
        diff[tin[v]] += vv
        diff[tout[v] + 1] -= vv

    # Khởi tạo BIT trong O(n) từ mảng hiệu (dựng cây Fenwick trực tiếp)
    bit = [0] * (size + 1)
    for i in range(1, size + 1):
        bit[i] += diff[i]
        j = i + (i & -i)
        if j <= size:
            bit[j] += bit[i]

    out = []
    for _ in range(q):
        t = data[pos]; pos += 1
        if t == b'2':
            # Truy vấn: tổng đường đi gốc -> s = prefix_sum(tin[s]) (inline vòng BIT)
            s = int(data[pos]); pos += 1
            i = tin[s]
            ssum = 0
            while i > 0:
                ssum += bit[i]
                i -= i & -i
            out.append(ssum)
        else:  # b'1'
            # Cập nhật giá trị đỉnh s: cộng delta tại tin[s], -delta tại tout[s]+1
            s = int(data[pos]); pos += 1
            x = int(data[pos]); pos += 1
            delta = x - val[s]
            val[s] = x
            i = tin[s]
            while i <= size:
                bit[i] += delta
                i += i & -i
            i = tout[s] + 1
            nd = -delta
            while i <= size:
                bit[i] += nd
                i += i & -i

    sys.stdout.buffer.write(('\n'.join(map(str, out)) + ('\n' if out else '')).encode())


main()

File: cses/06_tree_algorithms/1138_path-queries/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class FenwickTree {
public:
    explicit FenwickTree(int n) : bit_(n + 1) {}

    void add(int index, long long delta) {
        for (; index < static_cast<int>(bit_.size()); index += index & -index) {
            bit_[index] += delta;
        }
    }

    long long prefix_sum(int index) const {
        long long result = 0;
        for (; index > 0; index -= index & -index) {
            result += bit_[index];
        }
        return result;
    }

private:
    vector<long long> bit_;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, query_count;
    cin >> n >> query_count;
    vector<long long> value(n + 1);
    for (int node = 1; node <= n; ++node) {
        cin >> value[node];
    }

    // Danh sách kề của cây
    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < n - 1; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    // Euler tour bằng DFS lặp: gán entry[v] (lúc vào) và exit[v] (lúc ra) cho mỗi đỉnh.
    // child_index[node] ghi nhớ đã duyệt tới con thứ mấy của node.
    vector<int> parent(n + 1, -1), child_index(n + 1);
    vector<int> entry(n + 1), exit(n + 1);
    vector<int> stack = {1};
    parent[1] = 0;
    int timer = 1;
    entry[1] = timer;
    while (!stack.empty()) {
        int node = stack.back();
        if (child_index[node] < static_cast<int>(graph[node].size())) {
            int next = graph[node][child_index[node]++];
            if (next == parent[node]) {
                continue;  // Bỏ qua cạnh quay về cha
            }
            parent[next] = node;
            entry[next] = ++timer;  // Vào đỉnh con -> tăng timer
            stack.push_back(next);
        } else {
            exit[node] = timer;  // Ra khỏi node -> chốt exit
            stack.pop_back();
        }
    }

    // Mỗi đỉnh đóng góp trên đoạn Euler mà cây con của nó chiếm giữ.
    // Range-update được biểu diễn bằng hai point update trên BIT hiệu (difference):
    // +value tại entry[v], -value tại exit[v]+1.
    FenwickTree fenwick(n + 1);
    for (int node = 1; node <= n; ++node) {
        fenwick.add(entry[node], value[node]);
        fenwick.add(exit[node] + 1, -value[node]);
    }

    while (query_count--) {
        int type, node;
        cin >> type >> node;
        if (type == 1) {
            // Cập nhật giá trị đỉnh: chỉ cần cộng delta vào hai đầu đoạn của đỉnh
            long long new_value;
            cin >> new_value;
            long long delta = new_value - value[node];
            value[node] = new_value;
            fenwick.add(entry[node], delta);
            fenwick.add(exit[node] + 1, -delta);
        } else {
            // Tổng đường đi gốc -> node = tổng tiền tố tới entry[node]
            cout << fenwick.prefix_sum(entry[node]) << '\n';
        }
    }
}
